File: Python/Problems/Problems/problem316.py
import sys
from typing import List

from .problem import Problem


class Problem316(Problem):
    def test(self) -> bool:
        s = "bcabc"

        expected = "abc"

        result = self.removeDuplicateLetters(s)

        print(result)

        return result == expected

    def removeDuplicateLetters(self, s: str) -> str:
        n = len(s)
        indices = [sys.maxsize] * 26
        for i in range(n):
            indices[ord(s[i]) - ord('a')] = i
        stack = []
        inStack = [False] * 26

        for i in range(n):
            if inStack[ord(s[i]) - ord('a')]:
                continue

            while stack and ord(stack[-1]) > ord(s[i]) and i < indices[ord(stack[-1]) - ord('a')]:
                inStack[ord(stack[-1]) - ord('a')] = False
                stack.pop()

            inStack[ord(s[i]) - ord('a')] = True
            stack.append(s[i])


        return ''.join(stack)

File: C++/Problems/Problems/problem_316.cpp
#include "problem_316.h"

#include <stack>

bool problem_316::test() {
    std::string s = "bcabc";

    std::string expected = "abc";

    auto result = removeDuplicateLetters(s);

    std::cout << result << std::endl;

    return result == expected;
}

std::string problem_316::removeDuplicateLetters(std::string s) {
    int n = s.size();
    std::vector indices(26, INT_MAX);
    for (int i = 0; i < n; ++i)
        indices[s[i] - 'a'] = i;
    std::stack<char> stack;
    std::vector inStack(26, false);

    for (int i = 0; i < n; ++i)
    {
        if (inStack[s[i] - 'a'])
            continue;

        while (!stack.empty() && stack.top() > s[i] && i < indices[stack.top() - 'a'])
        {
            inStack[stack.top() - 'a'] = false;
            stack.pop();
        }

        inStack[s[i] - 'a'] = true;
        stack.push(s[i]);
    }

    std::string result = "";
    while (!stack.empty()) {
        result = stack.top() + result;
        stack.pop();
    }

    return result;
}
